algorithms/src/snark/marlin/ahp/matrices.rs

```diff
@@ -40,21 +40,39 @@ pub(crate) fn to_matrix_helper<F: Field>(matrix: &[Vec<(F, VarIndex)>], num_inpu
     cfg_iter!(matrix)
         .map(|row| {
             let mut row_map = BTreeMap::new();
-            row.iter().for_each(|(fe, column)| {
-                let column = match column {
-                    VarIndex::Public(i) => *i,
-                    VarIndex::Private(i) => num_input_variables + i,
-                };
-                *row_map.entry(column).or_insert_with(F::zero) += *fe;
+            row.iter().for_each(|(val, column)| {
+                if !val.is_zero() {
+                    let column = match column {
+                        VarIndex::Public(i) => *i,
+                        VarIndex::Private(i) => num_input_variables + i,
+                    };
+                    *row_map.entry(column).or_insert_with(F::zero) += *val;
+                }
             });
-            row_map.into_iter().map(|(column, coeff)| (coeff, column)).collect()
+            row_map.into_iter().map(|(column, val)| (val, column)).collect()
         })
         .collect()
 }
 
-/// This must *always* be in sync with `make_matrices_square`.
-pub(crate) fn padded_matrix_dim(num_formatted_variables: usize, num_constraints: usize) -> usize {
-    core::cmp::max(num_formatted_variables, num_constraints)
+/// Adds variables to randomize each z_M and preserve zero-knowledge
+/// When no random assignments are passed, we use F::one()
+pub(crate) fn add_randomizing_variables<F: PrimeField, CS: ConstraintSystem<F>, MM: MarlinMode>(
+    cs: &mut CS,
+    rand_assignments: Option<[F; 3]>,
+) {
+    let mut assignments = [F::one(); 3];
+    if MM::ZK {
+        if let Some(r) = rand_assignments {
+            assignments = r;
+        }
+    }
+
+    let zk_vars = assignments
+        .into_iter()
+        .enumerate()
+        .map(|(i, assignment)| cs.alloc(|| format!("random_{i}"), || Ok(assignment)).unwrap())
+        .collect_vec();
+    cs.enforce(|| "constraint zk", |lc| lc + zk_vars[0], |lc| lc + zk_vars[1], |lc| lc + zk_vars[2]);
```

Comment:
if MM::ZK is false, we should just not add new variables or the new constraint.

Reply:
1639d360f and 9071a88e7

Now if we don't have zero-knowledge, we hit my previously mentioned edge case: the degree of h_0.degree() is 0 when num_constraints == constraint_domain.size()
The first test which fails with prove_and_verify_with_square_matrix has properties: MM::ZK: false, mul_depth: 8, num_constraints: 32, num_variables: 32
The check fails because the only remaining unevaluated polynomial in the rowcheck equation is `L_row(k)^R(X)`, which becomes the degree 0 polynomial of `1` in this case because we use every row exactly once.

Do you think this might be a problem? The only way I see this edge case to be avoided is if we construct a dummy constraint if every row is used exactly once in one or all matrices. The downside is that this doubles the constraint_domain.size()

Reply:
I think it's fine. We're unlikely to hit that edge case in practice, but maybe we should add a test for it.

Reply:
Cool. It is indeed currently "triggered" by one of our existing test cases. Thank you for the quick response.